### PR TITLE
[NTUSER] Fix window not updating when scrollbars created.

### PR DESCRIPTION
**Pull request by Doug-Lyons**
I_Kill_Bugs patch.

## Purpose

_Fix text overwritten and not updated by scrollbars when scrollbars created._

JIRA issue: [CORE-19669](https://jira.reactos.org/browse/CORE-19669)

## Proposed changes

_Check winpos.c for change in WVR_HREDRAW or WVR_VREDRAW in function co_WinPosSetWindowPos._
_If one occurs then redraw full window._

Testbot Results:
winpos_min.patch JID68658 on top of 0.4.15-dev-8531-gd5f6b8c

KVM:  https://reactos.org/testman/compare.php?ids=97159,97165
VBox: https://reactos.org/testman/compare.php?ids=97162,97164

### CODE REVIEW

File: win32ss/user/ntuser/winpos.c
Comment (SigmaTel71) [inline win32ss/user/ntuser/winpos.c:1944]:
```suggestion
   if (WvrFlags & (WVR_HREDRAW | WVR_VREDRAW))
       fullRedraw = TRUE;
```

**Review by SigmaTel71 — APPROVED** [review]

File: win32ss/user/ntuser/winpos.c
Comment (JoachimHenze) [inline win32ss/user/ntuser/winpos.c:1940]: I will approve the PR, but allow me a question:
WVR_HREDRAW I do assume stands for horizontal redraw.
WVR_VREDRAW I do assume stands for vertical redraw.
What does WVR_REDRAW (from line 2056) stand for? Is that a combination of both?

**Review by JoachimHenze — APPROVED** [review]
Approved, but I would find it awesome if an automated testcase could be added to the bots which would cover this case. It took us many years to find it by hand. So potential future unintended breakage wouldn't be recognized early either.

File: win32ss/user/ntuser/winpos.c
Comment (JoachimHenze) [reply win32ss/user/ntuser/winpos.c:1940]: Yes, it is a combination, the constant WVR_REDRAW is defined as (WVR_HREDRAW | WVR_VREDRAW).
According to "Winconst tool (search Windows Constants)" http://oehive.org/book/export/html/466.html

That makes me wonder whether the additional local variable could be optimized away by expanding the condition in line 2056 instead.

**Review by oleg-dubinskiy — APPROVED** [review]

**Review by JoachimHenze — APPROVED** [review]
Renewing my approval, as requested. Adding Automated tests for that usecase would be great.

**Comment by dusxmt** [comment]
For an automated testcase, one idea I have is to deterministically change the color of a window whenever WM_SIZE is issued to it, with a key handler that would show or hide the scroll bar(s).  There could be a separate testcase for CS_HREDRAW and CS_WREDRAW.

The drawn color of the window could then be sampled, and if the wrong (old) color is retrieved, we know that WM_PAINT wasn't generated, and that the screen contents are out-of-date.

**Comment by dusxmt** [comment]
I've spent a bit of time putting together an automated testcase for this bug: https://jira.reactos.org/browse/ROSTESTS-397

File: win32ss/user/ntuser/winpos.c
Comment (HBelusca) [inline win32ss/user/ntuser/winpos.c:2050]: Two comments:
1. The second opening parenthesis (and the previous-to-last one at line 2054) are redundant and can be removed;

File: win32ss/user/ntuser/winpos.c
Comment (HBelusca) [inline win32ss/user/ntuser/winpos.c:2053]: Since now the test condition has been modified so that we _**require**_ always that `!(WvrFlags & WVR_REDRAW)` be true, this whole other line 2053 (and so, the `WinPos.flags & SWP_NOSIZE` test) now appears to be redundant.
Indeed, because `!(WvrFlags & WVR_REDRAW)` is now going to be always true, `(anything || TRUE)` will always be true.

File: win32ss/user/ntuser/winpos.c
Comment (Doug-Lyons) [reply win32ss/user/ntuser/winpos.c]: Removed extra parenthesis. Thanks.

File: win32ss/user/ntuser/winpos.c
Comment (Doug-Lyons) [reply win32ss/user/ntuser/winpos.c]: Simplified logic based on comments. Thanks.

**Comment by JoachimHenze** [comment]
Let's wait with merging the fix, until the test from 7254 has been merged already.

**Comment by mega-maniac** [comment]
If (WvrFlags & WVR_REDRAW) = True, there will be a memory leak of VisBeforeJustClient.

**Comment by Doug-Lyons** [comment]
Thanks @I_Kill_Bugs for the patch and approvals from @oleg-dubinskiy, @JoachimHenze, and @SigmaTel71.

**Comment by Doug-Lyons** [comment]
> If (WvrFlags & WVR_REDRAW) = True, there will be a memory leak of VisBeforeJustClient.

@mega-maniac, Do you have a suggested patch for this? Maybe you should create a Jira ticket for it. Thanks.

**Comment by dusxmt** [comment]
Thanks for reporting the memory leak issue @mega-maniac, I've opened https://github.com/reactos/reactos/pull/7284 to address it

**Pull request merged**
